### PR TITLE
In-house translation library implementation to support loading multiple MO files

### PR DESCRIPTION
**Pull request by BrettDong**
#### Summary
Infrastructure "In-house translation library implementation to support loading multiple MO files"

#### Purpose of change
The game uses `gettext` system for localisation. During run time, translated text is displayed on user interface by looking up the translation data loaded from `.mo` file on the disk. `.mo` (_Message Object_) is a translation file format used in `gettext` system, that basically stores a map from English text entries to their translated text entries, in a binary format which facilitates computer programs to process, but is not human readable. Human translators work on human-readable `.po` file format, and `.po` file is compiled to `.mo` during build time.

Currently, the game ships with a monolithic `cataclysm-dda.mo` (one for each translated language) that stores translations for all core game content and mainline mods (e.g. Aftershock, Magiclysm, etc.). When users launch the game, this `cataclysm-dda.mo` is loaded by `libintl`, the runtime component of `gettext` system (which is part of `glibc` on Linux, but is a standalone library on Windows), so these mainline game content are localised.

However, because of the way `gettext` system was designed, without making use of _domain_ (which the game definitely does not), the program can only use one `.mo` file at a time for translation, and hence rendering 3rd party mods translations impossible in the current framework. (#25566)

The current approach adopted in the community is to directly hack into mods JSON and replace all strings with translated text. https://github.com/linonetwo/CDDA-Kenan-Modpack-Chinese is such an example. This not only is a tedious process for translators and translation maintainers, but also does not work properly for languages like Russian, which has four plural forms, but the game's JSON API is designed for English and only supports a singular (`n == 1`) and a plural (`n != 1`) form.

There is also request of breaking down the current monolithic translation template into separate smaller translation files (e.g. GUI, Core Content, Mainline Mods) to ease translators navigating and prioritising the huge number of text entries. But it is technically infeasible due to the aforementioned limitation. (#23434)

#### Describe the solution
This pull request removes dependency on `libintl` and implements a built-in translation library that exposes similar interface to `libintl` but supports loading and using multiple `.mo` files at the same time.

`TranslationManager` singleton class is the external facing interface of the built-in translation library, that supports setting game language, and provides translation service. To reduce compile-time, the frequently included header file `translation_manager.h` only contains an interface, and the real implementation resides in `translation_manager_impl.{h, cpp}`.

A `TranslationDocument` class instance refers to a `.mo` file on the disk. The class reads a `.mo` file into memory and contains the translation data.

To support using multiple `.mo` files simultaneously, `TranslationManager` class maintains a vector of `TranslationDocument` class, and on language initialisation, computes a hash map from English string to a handle that records information on which location in which `TranslationDocument` to look into. A negative effect is that this will use slightly more run-time memory (26MB for Russian translation), but in return reduces look-up time (from `libintl`'s around 34 nanoseconds to around 16 nanoseconds, on Apple M1).

A tricky part is to support plural forms of different languages. In a `.mo` file, there is a C syntax math expression that describes the plural forms rules of the language. The expression has a variable `n`, which is the quantity, and plugging `n` into the expression gives the index of which plural form to use for this quantity. For example, for English, the singular form is stored at index `[0]` and the plural form is stored at index `[1]`. The expression for English is simply `n != 1`. When `n==1`, the expression evaluates to `0`, and the string at `[0]` is used, which is the singular form. When `n` is not `1`, the expression evaluates to `1`, and the string at `[1]` is used, which is the plural form. For Russian, the expression is much more complex, and it may evaluate to `0`, `1`, `2` or `3`.

`TranslationPluralRulesEvaluator` class is built for this purpose. It first performs lexical analysis on the input expression in string form to transform it into a vector of tokens, and then constructs an expression tree based on the sequence of tokens. With the expression tree, it's possible to evaluate the expression with any input quantity `n`.

Some test cases are added to cover the functionalities of these three classes.

To avoid symbol collision with `libintl`, all occurrences of `ngettext` are renamed to `n_gettext`.

#### Describe alternatives you've considered
An alternative approach is to merge `.mo` files during run-time, and keep using `libintl` for run-time translation.

1. This still requires implementing a `.mo` file reader, and after merging the text entries, furthermore, a `.mo` file writer is also required.
2. Merging `.mo` files could take time and cause user sensible delay on start up.

#### Testing

Launch game and expect localisation to function normally.

* CI passing
* Tested on macOS, no sensible difference in game play
* Tested on Windows

Compiled binaries for testing: https://github.com/BrettDong/Cataclysm-DDA/releases/tag/cdda-experimental-2021-09-15-1625

#### Benchmark
Test Environment: Apple M1 MacBook Pro, macOS 11.5.2, Apple Clang 12.0.5.
Build parameters: `make RELEASE=1 TILES=1 -j8`
Measure time spent on `gettext()` and `ngettext()` for `libintl`.

| Game Language set to English | `libintl` | TranslationManager |
|---|---|---|
| Translate singular "battery" | 6,654 ns | 12 ns |
| Translate plural "battery" | 6,679 ns | 12 ns |

| Game Language set to Russian | `libintl` | TranslationManager |
|---|---|---|
| Translate singular "battery" | 33 ns | 16 ns |
| Translate plural "battery" | 66 ns | 50 ns |

| Game Language set to Simplified Chinese | `libintl` | TranslationManager |
|---|---|---|
| Translate singular "battery" | 33 ns | 16 ns |
| Translate plural "battery" | 35 ns | 19 ns |

#### Memory usage
Launch the program and exit from title menu. Observe "Maximum resident set size".

| Language |  `libintl` | TranslationManager |
|---|---|---|
| English | 87,536 KB | 87,088 KB |
| Russian | 88,976 KB | 115,824 KB |
| Simplified Chinese | 88,704 KB | 102,912 KB |

The benchmark above indicates that this pull request does not cause performance regression, but is even faster than `libintl`, however memory usage is slightly higher.

This benchmark also confirms that there is indeed a significant performance penalty on `libintl` if translation is not found in `.mo` file (e.g. game language is set to English), not only on MinGW (#38849), but also on other platforms including macOS. The translation library implementation in this pull request does not have such problem.

#### Impact

Benefits:
* Technical foundation for loading separate `.mo` files (#23434)
* Modular 3rd part mods translation is now possible (#25566, #44225)
* No longer need to set environment variables (#47466, #48176 can retire) and avoids related pitfalls on Windows (#45238)
* No longer need `libintl-lite` hack for Android (hacks like #45795, and #48354 can retire) and avoids related problems (#48349)
* No longer depends on `libintl` and may help with i18n on `musl` C library (#39673)
* Solves performance problem when game language is set to English (#38849) and the workaround of compiling a dummy English `.mo` translation is no longer necessary (#38892, #46147, #47358 can retire)

Disadvantages:
* Slightly higher memory usage (<30MB)

#### Additional context
To code reviewers: reviewing each commit individually might be easier to follow.

### CODE REVIEW

**Comment by andrei8l** [comment]
There's still no built-in support for grammatical gender. Should that be done with contexts for this library as well? That seemed like a bit of a hack in gettext.

**Comment by Qrox** [comment]
Based on my understanding, we still need to import the translation files on Transifex, so unless we switch to some other format that supports grammatical gender, we still have to hack it using context.

**Comment by BrettDong** [comment]
Grammatical gender is not supported by .mo format, so even an in house reimplementation like this pull request cannot deliver native support for grammatical gender, unless we completely switch to another translation file format that supports grammatical gender and also is supported by Transifex.

**Comment by andrei8l** [comment]
You can do some internal processing and still have valid `.mo` files. For example, you could prefix the message with a character representing gender (and no prefix means no gender so it's backwards-compatible). That's still kinda hacky, but leaves the context free for its intended uses.

**Comment by BrettDong** [comment]
That's intrusive to the message body and I cannot accept it.

**Comment by andrei8l** [comment]
It's only intrusive to the `msgid`, not `msgstr`. I've written a quick demo [here](https://github.com/andrei8l/Cataclysm-DDA/commit/8878650a9195b2670892569cc43809ecc96fdc60). However, you cannot use `xgettext` for this.

**Comment by BrettDong** [comment]
Let's move the topic of grammatical gender to a separate issue or pull request. Currently I don't plan to implement grammatical gender in this pull request.

File: data/mods/TEST_DATA/lang/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+mo/
```
Comment (Qrox) [inline data/mods/TEST_DATA/lang/.gitignore:1]: Maybe exclude `INVALID_RAND.mo` from .gitignore?

Comment (Qrox) [reply data/mods/TEST_DATA/lang/.gitignore:1]: What does git do when someone else modifies the added but ignored .mo files? Do they show up in the modified files list? If not people might forget to commit them when they are changed.

Comment (BrettDong) [reply data/mods/TEST_DATA/lang/.gitignore:1]: It does get listed:
```
[brett] ~/Cataclysm-DDA [mo_loader]  % vim data/mods/TEST_DATA/lang/mo/ru/LC_MESSAGES/INVALID_RAND.mo 
[brett] ~/Cataclysm-DDA [mo_loader] M % git status
On branch mo_loader
Your branch is up to date with 'brett/mo_loader'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   data/mods/TEST_DATA/lang/mo/ru/LC_MESSAGES/INVALID_RAND.mo

no changes added to commit (use "git add" and/or "git commit -a")
```

**Comment by actual-nh** [comment]
Question: Why would merging the .mo files need to be during run-time and not build-time?

**Comment by BrettDong** [comment]
> Question: Why would merging the .mo files need to be during run-time and not build-time?

The motivation is to support modular 3rd party mods translations, and users may add or remove 3rd party mods from time to time. In the alternative approach, the program will need to, on every start up, validate cache and rebuild (re-merge) `.mo` if necessary.

**Comment by actual-nh** [comment]
> > Question: Why would merging the .mo files need to be during run-time and not build-time?
> 
> The motivation is to support modular 3rd party mods translations, and users may add or remove 3rd party mods from time to time. In the alternative approach, the program will need to, on every start up, validate cache and rebuild (re-merge) `.mo` if necessary.

Ah. I'd thought the intent was to be able to split up vanilla from in-repo mod translations, but good point re 3rd-party mods.

**Comment by jbytheway** [comment]
I think Bright Nights made a similar change.  Does yours relate to theirs in any way?

**Comment by BrettDong** [comment]
> I think Bright Nights made a similar change. Does yours relate to theirs in any way?

From description of https://github.com/cataclysmbnteam/Cataclysm-BN/pull/505 I think they picked the same approach to support modular 3rd party mods translations, but the code implementation is different.

**Comment by jbytheway** [comment]
I think this is what I was thinking of: https://github.com/cataclysmbnteam/Cataclysm-BN/pull/505

**Comment by Ezsk** [comment]
Is there any way for generating .po files from the mod? Maybe like a python tool or something.
 As a modder, I'm very interested in this feature, but I don't know how to do it specifically. I guess documentation and guides are needed.

I apologize if this is not the suitable place to ask this question.

**Comment by ZhilkinSerg** [comment]
Everything is in lang folder in the repo.

**Comment by BrettDong** [comment]
A formal template generation script is not available yet, but you can modify the Python script in `lang` folder to cater to your needs.

**Comment by Ezsk** [comment]
Thank you for the response. However, it seems a difficult task for the ordinary modder and translation contributors. I will try to check lang folder, but it will take quite a while before we can use the feature.

**Comment by ZhilkinSerg** [comment]
Shall we remove gettext references from everywhere now (appveyor, vcpkg, etc)?

**Comment by ZhilkinSerg** [comment]
Do we really need both checks in here?

https://github.com/CleverRaven/Cataclysm-DDA/blob/630e438b9b40691db26a2b28ef640b36715bd965/src/translation_plural_evaluator.cpp#L337

**Comment by BrettDong** [comment]
> Do we really need both checks in here?
> 
> 
> 
> https://github.com/CleverRaven/Cataclysm-DDA/blob/630e438b9b40691db26a2b28ef640b36715bd965/src/translation_plural_evaluator.cpp#L337

Must be a typo, the latter condition should check the size of operators vector.

**Pull request merged**